rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+that does not carry behavior (for now).
+
+```sway
+trait AbiError: AbiEncode {}
+```
+
+We need to introduce two new annotations.
+
+`#[abi_error]` generates an implementation of `AbiError` for a given type
+and should let the compiler know that it should carry around error message
+information about this type.
+
+`#[error("")]` should only be usable in a `#[abi_error]` marked enum definition,
+and to not exhaustively mark variants of such an enum with an error message
+annotation is considered an error.
+
+When generating ABI files, the compiler should populate the `"metadataTypes"`
+section accordingly and include a `"errorMessage"` field in the components of
+marked enums.
+
+## `panic` intrinsic
+
+The `panic` intrinsic has two modes which determine what kind of fields are
+produced in the (new) `"errorCodes"` section of the ABI file.
+
+Either it is used with a `str` litteral (or a const-evaluable such value) and
+produces `"msg"` fields. Or it is used with an `AbiError` implementing type, and
+produces `"logId"` fields. Any other use is an error.
+
+In any valid case, it also records in the `"errorCodes"` section location
+information about every `panic` invocation's position in the codebase.
+
+As for runtime behavior, if the argument is an error type, it logs it before
+reverting, and in every valid case, it eventually reverts with a unique error
+code for every `panic` invocation, which is recorded in the `"errorCodes"`
+section.
+
+## Integration considerations
+
+Manual reverts with custom codes should still be available to developers, but
+discouraged.
+
+The standard library should henceforth strive to use error types for both
+recoverable and irrecoverable errors. Ideally the standard library should
+never produce a revert code that doesn't have a corresponding documented error
+message.
+
+We should also introduce compiler warnings for manually producing revert codes
+that may conflict with either the reserved code range or the autogenerated panic
+code range.
+
+SDK integration of this feature is open ended, but we should at least aim to
+be able to use error message information to decode revert codes and error types
+that are directly returned.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+Using a special intrinsic is the main drawback of this solution for a number of reasons.
+
+It muddles how instrinsics are identified since unlike other compiler intrinsics
+`panic` does not have a `__` prefix. 
+This is necessary since it needs to
+be directly invoked at the error site for the location information to be
+meaningful.
+
+It also pollutes the namespace with yet another reserved word.
+
+And ultimately it moves more complexity to the compiler rather than let the user define it.
+This is a shared problem with the error type annotations.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Despite the drawbacks of moving more special behavior the compiler, it has to
```

Comment:
```suggestion
Despite moving more special behavior to the compiler, Sway must shoulder this complexity since it lacks sufficiently powerful metaprogramming features that would allow developers to define their own language features of this caliber.
```